### PR TITLE
[Merged by Bors] - feat: better display of partial results from library_search

### PR DESCRIPTION
**Pull request by kim-em**
On @PatrickMassot's [example](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/library_search.20regression/near/353375130):
```lean
import Mathlib.Topology.Instances.Real
import Mathlib.Topology.Algebra.Order.Compact
import Mathlib.Tactic.LibrarySearch

example (f : ℝ → ℝ) {K : Set ℝ} (hK: IsCompact K) : ∃ x ∈ K, ∀ y ∈ K, f x ≤ f y := by
  library_search
```
we have:
* `refine IsCompact.exists_forall_le hK ?_ ?_` as one of the suggested solutions
* in fact, as the first solution (on the basis that it uses more local hypotheses than alternatives)
* none of the spurious results which use `False.elim ?_`
* and better display of the result, with hints:

```lean
refine IsCompact.exists_forall_le hK ?_ ?_
-- Remaining subgoals:
-- ⊢ Set.Nonempty K
-- ⊢ ContinuousOn (fun x ↦ f x) K
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Lean/Meta.lean [hunk 1]
```diff
@@ -83,17 +84,55 @@ partial def intros! (mvarId : MVarId) : MetaM (Array FVarId × MVarId) :=
...
+Check if a goal is "independent" of a list of other goals.
+We say a goal is independent of other goals if assigning a value to it
+can not change the solvability of the other goals.
+
+This function only calculates an approximation of this condition
+-/
+def independent? (L : List MVarId) (g : MVarId) : MetaM Bool := do
+  let t ← instantiateMVars (← g.getType)
+  if t.hasExprMVar then
+    -- The the goal's type contains other meta-variables,
```
Comment (semorrison) [inline Mathlib/Lean/Meta.lean:105]:
```suggestion
    -- If the goal's type contains other meta-variables,
```

File: Mathlib/Tactic/SolveByElim.lean [hunk 2]
```diff
@@ -213,8 +213,8 @@ By default `cfg.suspend` is `false,` `cfg.discharge` fails, and `cfg.failAtMaxDe
 and so the returned list is always empty.
 Custom wrappers (e.g. `apply_assumption` and `apply_rules`) may modify this behaviour.
 -/
-def solveByElim (cfg : Config) (lemmas : List (TermElabM Expr)) (ctx : TermElabM (List Expr))
-    (goals : List MVarId) : MetaM (List MVarId) := do
+partial def solveByElim (cfg : Config) (lemmas : List (TermElabM Expr))
+    (ctx : TermElabM (List Expr)) (goals : List MVarId) : MetaM (List MVarId) := do
```
Comment (semorrison) [inline Mathlib/Tactic/SolveByElim.lean:217]:
```suggestion
def solveByElim (cfg : Config) (lemmas : List (TermElabM Expr)) (ctx : TermElabM (List Expr)) 
    (goals : List MVarId) : MetaM (List MVarId) := do
```

**Comment by semorrison** [comment]
bors merge

**Comment by PatrickMassot** [comment]
bors merge

**Pull request closed without merging**